### ISSUE TITLE
[java] UnusedImports false positive for wildcard import

### ISSUE THREAD

**Issue by maikelsteneker** [pmd/pmd#1222]
**Affects PMD Version:**

6.5.0 and 6.4.0. It was not present in PMD 5.3.0.

**Rule:**

category/java/bestpractices.xml/UnusedImports

**Description:**

When using a wildcard import (*), PMD sometimes reports false positives. These are imports that it claims can be removed, but when removing them the code cannot compile anymore. Replacing the wildcard by a single class import solves the issue.

**Code Sample demonstrating the issue:**

```
import com.intellij.openapi.actionSystem.*;

public class MyClass extends com.intellij.openapi.actionSystem.AnAction {
    @Override
    public void actionPerformed(AnActionEvent event) {
    }
}
```

**Running PMD through:** command line: `./run.sh pmd -dir /tmp/pmd_fp/src/main/java/ -rulesets /tmp/pmd_fp/pmd.xml`
[Gradle project to reproduce the issue](https://github.com/pmd/pmd/files/2158639/pmd_fp.zip)

**Comment by jsotuyod** [comment]
@maikelsteneker this is down to you not passing the `auxclasspath` to PMD, so type resolution can't take place. If you run your own example with the gradle task (`./gradlew pmdMain`) the violation is not reported.

We could ignore all star imports upon missing classes, but that would just convert this noisy false positive into silent (and dangerous) false negatives. I'd therefore avoid it, at least until #194 is done and shipped, having all missing classes reported.

**Comment by maikelsteneker** [comment]
@jsotuyod Thanks for the insight, I didn't realize that that is what caused the issue! When I provide the proper classpath, the violation is indeed no longer reported.

I think reporting errors as suggested in #194 would be extremely helpful for cases like these, because to inexperienced users, it's not obvious where a problem like this comes from.

**Comment by Fernal73** [comment]
This rule could be relaxed in the case of static imports allowing the programmer to specify a static import *, as long as the number of unused methods or fields imported don't exceed a configurable limit.

**Comment by Fernal73** [comment]
Is this going to be looked at? It can become very annoying when because one or two methods from a static import not used forces this warning to be displayed. Yes, it can be suppressed but with interfaces no longer supporting Constants or rather, coding standards discouraging the use of contracts for the said purpose, it becomes necessary to allow a relaxation of this rule.

**Comment by adangel** [comment]
@Fernal73 If you think, this rule could be improved, please create a new issue with your suggestions and your use case. You are commenting here a closed issue - closed by definition means no work/time will be invested here. This issue is done - that's why it's closed. Thanks!

**Comment by Fernal73** [comment]
OK. I was hoping that you'd tell me if it warranted raising an issue.

Here goes nothing:

https://github.com/pmd/pmd/issues/1870

Regards,
Linus.
